sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,314 @@
+ */
+
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       STDIO over UART + NimBLE implementation
+ *
+ * This file is based on stdio_uart.c and is extended by functionality to access
+ * stdio via NimBLE. This module is designed to use UART and NimBLE in parallel.
+ *
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "stdio_uart.h"
+#include "stdio_nimble.h"
+
+#include "periph/uart.h"
+#include "isrpipe.h"
+
+#include "ringbuffer.h"
+
+#define MTU_MAX_NIMBLE BLE_ATT_MTU_MAX
+#define STDIN_BUFFER_SIZE MTU_MAX_NIMBLE
+#define RINGBUFFER_SIZE MTU_MAX_NIMBLE
+
+#define ATT_PKT_HEADER_SIZE 13
+
+#define MAX_READ (MTU_MAX_NIMBLE - ATT_PKT_HEADER_SIZE)
+
+/* isrpipe for stdin */
+static uint8_t _isrpipe_mem[STDIO_UART_BUFSIZE];
+static isrpipe_t _stdin_isrpipe = ISRPIPE_INIT(_isrpipe_mem);
+
+/* simple buffer for stdin */
+static char _stdin_data[STDIN_BUFFER_SIZE];
+
+/* ringbuffer for stdout */
+static char _rb_mem[RINGBUFFER_SIZE];
+static char _rb_read_buf[MAX_READ];
+static ringbuffer_t _rb;
+static mutex_t _rb_mutex = MUTEX_INIT;
+static int _current_max_read = MAX_READ;
+
+/* information about bluetooth connection */
+static uint16_t _conn_handle;
+static bool _send_stdio = false;
+static bool _sending_in_progress = false;
+static mutex_t _send_mutex = MUTEX_INIT;
+
+static struct ble_gap_event_listener _gap_event_listener;
+
+/**
+ * @brief Struct to define the stdio bluetooth service with its characteristics
+ */
+static const struct ble_gatt_svc_def _gatt_svr_svcs[] =
+{
+    /*
+     * access_cb defines a callback for read and write access events on
+     * given characteristics
+     */
+    {
+        /* Service: stdio */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = (ble_uuid_t*) &gatt_svr_svc_stdio_uuid.u,
+        .characteristics = (struct ble_gatt_chr_def[]) { {
+            /* Characteristic: stdout */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdout_uuid.u,
+            .access_cb = gatt_svr_chr_access_noop,
+            .val_handle = &_val_handle_stdout,
+            .flags = BLE_GATT_CHR_F_INDICATE,
+        }, {
+            /* Characteristic: stdin */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdin_uuid.u,
+            .access_cb = gatt_svr_chr_access_stdin,
+            .flags = BLE_GATT_CHR_F_WRITE,
+        }, {
+            0, /* No more characteristics in this service */
+        }, }
+    },
+    {
+        0, /* No more services */
+    },
+};
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg);
+
+static void _send_from_ringbuffer(bool mutex_locked);
+
+/* copy-pasted from stdio_uart module */
+#ifndef STDIO_NIMBLE_DISABLE_UART
+static void _init_stdio_uart(isrpipe_t *isrpipe)
+{
+    uart_rx_cb_t cb = (uart_rx_cb_t) isrpipe_write_one;
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, cb, isrpipe);
+}
+#endif
+
+static void _stdio_nimble_write(const void* buffer, size_t len)
```

Comment:
same here, there is not really a reason for splitting this function if it is only used once. For a bad compiler this might just increase code size...

Reply:
Because UART-in-parallel feature is removed, I merged `_stdio_nimble_write` in to `stdio_write`. I generally like it to put code for different purposes in to their own functions. Both are for writing data, but if you see it at a lower level one is for sending to uart and the other for sending to nimble. That's why I've put them in separated functions.